text/0000-stabilize-os-str-conversions.md

```diff
@@ -0,0 +1,139 @@
+
+# Detailed design
+
+First, functions for converting from an array of bytes into an OS string will be
+stabilized, the two signatures in question are:
+
+```rust
+impl OsString {
+    fn from_platform_bytes(bytes: Vec<u8>) -> Option<OsString>;
+}
+
+impl OsStr {
+    fn from_platform_bytes(bytes: &[u8]) -> Option<&OsStr>;
+}
+```
+
+> Note: the `OsString::from_bytes` function today has been renamed here and the
+> generics have been removed.
+
+The name contains "platform" to indicate that the function has platform specific
+behavior, and the semantics will be:
+
+* On Unix, return `Some(mem::transmute(bytes))`
+* On Windows, interpret the bytes as utf-8, returning `Some` if this succeeded
+  or `None` if it failed.
+
+Next, some convenience functions will be stabilized for extracting a list of
+bytes out of an OS string:
+
+```rust
+impl OsStr {
+    fn to_bytes(&self) -> Option<&[u8]>;
+    fn to_cstring(&self) -> Option<CString>;
+}
+```
+
+The semantics of these functions will be:
+
+* On Unix, return `Some` always on `to_bytes` and call `CString::new(..).ok()`
+  for `to_cstring`.
+* On Windows, attempt to interpret the `&[u16]` as utf-16, and if successful
+  convert to utf-8 and perform the same as Unix. If the utf-16 interpretation
+  fails then `None` is returned.
+
+# Drawbacks
+
+The platform-specific behavior of these functions can be surprising to some
+programs. It's relatively easy to leverage these functions and witness that they
+never fail in a Unix environment, allowing use of `unwrap` to go unnoticed and
+discouraging proper error handling. There is, however, a general culture of
+avoiding `unwrap` in robust code in Rust.
+
+Additionally, although these functions have platform-specific behavior, they're
+enabling as many successful use cases on each platform as possible. The failure
+mode of `u16` to `u8` conversion on Windows is the only case where `None` is
+returned, and it's inevitable that applications need to make a decision of what
+to do in this case regardless (e.g. apply a lossy conversion or return an
+error).
+
+# Alternatives
+
+Outlined in [rust-lang/rust#27657][pr], an alternative would be to provide
+infallible interpretations of OS strings as either `[u16]` or `[u8]`. In each
+direction either Unix or Windows would be lossless but the opposite platform
+would use the `from_utf8_lossy` family of functions on strings to replace
+ill-formed unicode with unicode replacement characters.
+
+[pr]: https://github.com/rust-lang/rust/pull/27657
+
+The downside of this approach, however, is that an infallible conversion implies
+some form of lossiness across platforms, which is easy to forget and start
+silently relying on in applications. Additionally, it's possible to build the
+lossy version on top of the non-lossy versions proposed in this RFC.
+
+# Unresolved questions
+
+* Is `platform` an appropriate name to have in these methods? Does it correctly
```

Comment:
When reading the summary I first thought that "platform bytes" on Windows meant UTF-16. Which had me worried because it might not be what one wants, and even if it is it depends on CPU endianness.